keyboards/kiwikey/kawii9v2/lib/oled.c

```diff
@@ -0,0 +1,242 @@
+                          (GRAPH_ZERO_Y + 1) / 8 - 1,
+                          false);
+        }
+        graph_timer = timer_read();
+    }
+}
+
+void ui_clear(void) {
+    oled_set_cursor(0, 4);
+    oled_advance_page(true);
+    oled_advance_page(true);
+    oled_advance_page(true);
+    oled_advance_page(true);
+}
+
+void render_ui_frame(void) {
+    oled_clear();
+    oled_advance_page(false);
+    oled_write_ln_P(PSTR("LAYER    WPM"), false);
+    oled_advance_page(false);
+    oled_write_ln_P(PSTR("RGB"), false);
+    oled_advance_page(false);
+    oled_advance_page(false);
+    oled_advance_page(false);
+    // Outline for layer number
+    draw_line_h(35, 7, 7, true);
+    draw_line_v(34, 8, 7, true);
+    draw_line_v(35, 8, 8, true);
+    draw_line_v(42, 8, 7, true);
+    // Outline for RGB mode number
+    draw_line_h(23, 23, 19, true);
+    draw_line_v(22, 24, 7, true);
+    draw_line_v(23, 24, 8, true);
+    draw_line_v(42, 24, 7, true);
+    // Outline for Key Matrix
+    draw_line_h(MATRIX_DISPLAY_X -2, MATRIX_DISPLAY_Y -3, MATRIX_DISPLAY_SIZE, true);
+    draw_line_h(MATRIX_DISPLAY_X -2, MATRIX_DISPLAY_Y -2 + MATRIX_DISPLAY_SIZE, MATRIX_DISPLAY_SIZE, true);
+    draw_line_v(MATRIX_DISPLAY_X -3, MATRIX_DISPLAY_Y -2, MATRIX_DISPLAY_SIZE, true);
+    draw_line_v(MATRIX_DISPLAY_X -2 + MATRIX_DISPLAY_SIZE, MATRIX_DISPLAY_Y-2, MATRIX_DISPLAY_SIZE, true);
+}
+
+void render_ui_rgbcontrol(void) {
+    oled_set_cursor(0, 5);
+    oled_write_P(PSTR("----< Lighting >-----"), false);
+    // RGB mode
+    oled_set_cursor(0, 6);
+    oled_write_ln(rgb_matrix_name(rgb_matrix_get_mode()), false);
+    // RGB brightness bar
+    oled_set_cursor(0, 7);
+    oled_write_char((rgb_val == 0) ? 0x07 : 0x9D, false);
+    for (uint8_t i = 0; i < 20; i++) {
+        oled_write_char(0x20, i*5 < rgb_val);
+    }
+}
+
+void render_stats(void) {
+    // Current layer
+    oled_set_cursor(6, 1);
+    oled_write_char(get_highest_layer(layer_state) + 0x30, true);
+    // Current WPM
+    oled_set_cursor(13, 1);
+    oled_write(get_u8_str(get_current_wpm(), '0'), false);
+    // Show current RGB mode (#__)
+    oled_set_cursor(4,3);
+    if (rgb_matrix_is_enabled()) {
+        oled_write_P(PSTR("#"), true);
+        oled_write_char(rgb_matrix_get_mode()/10 + 0x30, true);
+        oled_write_char(rgb_matrix_get_mode()%10 + 0x30, true);
+    }
+    else {
+        oled_write_P(PSTR("OFF"), true);
+    }
+    // RGB brightness (percentage %)
+    oled_set_cursor(9, 3);
+    rgb_val = (rgb_matrix_is_enabled() ? rgb_matrix_get_val()*100/RGB_MATRIX_MAXIMUM_BRIGHTNESS : 0);
+    if (rgb_matrix_is_enabled()) {
+        oled_write_char((rgb_val == 0) ? 0x07 : 0x9D, false);
+        oled_write(get_u8_str(rgb_val, ' '), false);
+        oled_write_char(0x25, false); // %
+    }
+    else {
+        oled_write_P(PSTR("-----"), false);
+    }
+}
+
+
+void keyboard_post_init_user(void) {
+    oled_on();
+    oled_init(OLED_ROTATION_0);
+    render_ui_frame();
+}
```

Comment:
`_user` functions at the keyboard level are invalid. The various instances in this file would need to be refactored to `_kb`. However there is then the issue of how a user would override the behaviour, which is not possible in some of the provided implementation.

Reply:
Hi @zvecr , thanks for your review on my PR.

I made changes as your recommendation (mostly about moving configurations from `config.h` to `info.json`, and default values/functions that no need to be implemented).
After changes, QMK MSYS compiled successfully and I also tested on actual hardware, all works fine (as I can manage).

Regarding the issue of how a user would override the behaviour of OLED screen, I have no idea for now (frankly I am an amateur in coding). My suggestion is we just keep these implementations at keyboard-level, and there should be a PR in future to resolve this.